File: src/generator/integer.ts
import { Arbitrary, Generator } from '../Generator'
import { Random } from '../Random'
import { generateInteger } from '../shrinker/integer'

/**
 * Generates integers within the specified inclusive range [min, max].
 *
 * @param min The minimum value of the range (inclusive). Defaults to `Number.MIN_SAFE_INTEGER`.
 * @param max The maximum value of the range (inclusive). Defaults to `Number.MAX_SAFE_INTEGER`.
 * @returns A generator for integers within the specified range.
 * @throws If min is greater than max.
 */
export function interval(min: number = Number.MIN_SAFE_INTEGER, max: number = Number.MAX_SAFE_INTEGER): Generator<number> {
    if (min > max) throw new Error(`invalid range: min (${min}) > max (${max})`)
    return new Arbitrary((random: Random) => {
        return generateInteger(random, min, max)
    })
}

/**
 * Generates integers within the specified range [fromInclusive, toExclusive).
 *
 * @param fromInclusive The minimum value of the range (inclusive).
 * @param toExclusive The maximum value of the range (exclusive).
 * @returns A generator for integers within the specified range.
 * @throws If fromInclusive is greater than or equal to toExclusive.
 */
export function inRange(fromInclusive: number, toExclusive: number): Generator<number> {
    if (fromInclusive >= toExclusive) throw new Error(`invalid range: from (${fromInclusive}) >= to (${toExclusive})`)
    return interval(fromInclusive, toExclusive - 1)
}

/**
 * Generates a sequence of `count` integers starting from `start`.
 * Equivalent to `inRange(start, start + count)`.
 * @deprecated Use `Gen.interval` or `Gen.inRange` instead.
 * @param start The starting integer (inclusive).
 * @param count The number of integers to generate.
 * @returns A generator for the sequence of integers.
 */
export function integers(start: number, count: number): Generator<number> {
    return inRange(start, start + count)
}

/**
 * Generates a random integer.
 * @returns A generator for a random integer.
 */
export function integer(): Generator<number> {
    return interval()
}

File: dartproptest/lib/src/generator/integer.dart
import '../generator.dart';
import '../shrinker/integer.dart';

/// Generates integers within the specified inclusive range [min, max].
///
/// [min] The minimum value of the range (inclusive). Defaults to safe integer bounds.
/// [max] The maximum value of the range (inclusive). Defaults to safe integer bounds.
/// Returns a generator for integers within the specified range.
/// Throws If min is greater than max.
Generator<int> interval([int? min, int? max]) {
  final minVal = min ?? -9007199254740991; // Number.MIN_SAFE_INTEGER equivalent
  final maxVal = max ?? 9007199254740991; // Number.MAX_SAFE_INTEGER equivalent

  if (minVal > maxVal)
    throw Exception('invalid range: min ($minVal) > max ($maxVal)');
  return Arbitrary((random) {
    return generateInteger(random, minVal, maxVal);
  });
}

/// Generates integers within the specified range [fromInclusive, toExclusive).
///
/// [fromInclusive] The minimum value of the range (inclusive).
/// [toExclusive] The maximum value of the range (exclusive).
/// Returns a generator for integers within the specified range.
/// Throws If fromInclusive is greater than or equal to toExclusive.
Generator<int> inRange(int fromInclusive, int toExclusive) {
  if (fromInclusive >= toExclusive) {
    throw Exception(
        'invalid range: from ($fromInclusive) >= to ($toExclusive)');
  }
  return interval(fromInclusive, toExclusive - 1);
}

/// Generates a sequence of [count] integers starting from [start].
/// Equivalent to `inRange(start, start + count)`.
///
/// @deprecated Use `interval` or `inRange` instead.
/// [start] The starting integer (inclusive).
/// [count] The number of integers to generate.
/// Returns a generator for the sequence of integers.
Generator<int> integers(int start, int count) {
  return inRange(start, start + count);
}

/// Generates a random integer.
/// Returns a generator for a random integer.
Generator<int> integer() {
  return interval();
}
